### PR TITLE
ci: Test output consistency between constant-folding vs. rendered query evaluations

### PR DESCRIPTION
**Pull request by nrainer-materialize**
This fixes https://github.com/MaterializeInc/materialize/issues/17767.

TODO:
* ✅ compare results
* ✅ create issues where necessary and link them (all issues already exist, no new problems found)

### CODE REVIEW

File: test/sqllogictest/transform/fold_constants.slt [hunk 1]
```diff
@@ -143,3 +143,650 @@ Explained Query:
           Get l0 // { arity: 1 }
 
 EOF
+
```
Comment (nrainer-materialize) [inline test/sqllogictest/transform/fold_constants.slt:146]: Shall I move the following code to separate files to reduce the length of this one?

File: test/sqllogictest/transform/fold_constants.slt [hunk 2]
```diff
@@ -143,3 +143,650 @@ Explained Query:
...
+# 4c) cleanup
+
+statement ok
+DROP VIEW v_using_constant_folding;
+
+statement ok
+DROP TABLE t_using_dataflow_rendering;
+
+# -------------------------------
+# 5) repeat row (=> this panics!)
```
Comment (nrainer-materialize) [inline test/sqllogictest/transform/fold_constants.slt:463]: This panics and lets the test crash. How to proceed here?
https://github.com/MaterializeInc/materialize/issues/17758#issuecomment-1438824467 mentions that it requires requires `--unsafe-mode`.

**Comment by nrainer-materialize** [comment]
> @nrainer-materialize did you write those queries by hand or you used a script to generate them?

Queries are handwritten, results are generated.

File: test/sqllogictest/transform/fold_constants.slt (hunk 1 above)
Comment (nrainer-materialize) [reply test/sqllogictest/transform/fold_constants.slt]: => I split them by computation mode (to allow easier comparisons) and content (to reduce the file length).

**Review by def- — CHANGES_REQUESTED** [review]
> [2023-04-18T08:59:54Z] lint: error: copyright: test/sqllogictest/transform/fold_vs_dataflow/1_numbers_dataflow.slt is missing copyright header

You can copy the copyright header from another file

File: test/sqllogictest/transform/fold_vs_dataflow/3_number_aggfns_dataflow.slt [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+query IIIIIIIIIIIIIIIIIIIIIIII
+SELECT
+ SUM(real1), SUM(double1), SUM(numeric1),
+ SUM(real1 + real1), SUM(double1 + double1), SUM(numeric1 + numeric1),
+ MIN(real1), MIN(double1), MIN(numeric1),
+ MIN(real1 + real1), MIN(double1 + double1), MIN(numeric1 + numeric1),
+ MAX(real1), MAX(double1), MAX(numeric1),
+ MAX(real1 + real1), MAX(double1 + double1), MAX(numeric1 + numeric1),
+ AVG(real1), AVG(double1), AVG(numeric1),
+ AVG(real1 + real1), AVG(double1 + double1), AVG(numeric1 + numeric1)
```
Comment (vmarcos) [inline test/sqllogictest/transform/fold_vs_dataflow/3_number_aggfns_dataflow.slt:37]: Should we have expressions here also with the integer columns?

File: test/sqllogictest/transform/fold_vs_dataflow/5_repeat_row_dataflow.slt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  SELECT data FROM t_base, repeat_row(diff);
+
+# see also: https://github.com/MaterializeInc/materialize/issues/17758
+# query error
+# SELECT SUM(data)
+# FROM mv_using_dataflow_rendering;
+# resulting in:
+# thread 'tokio-runtime-worker' panicked at 'cluster-u1-replica-1-0 crashed; aborting because propagate_crashes is enabled', src/orchestrator-process/src/lib.rs:560:29
+# note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+# 2023-04-18T09:50:00.705649Z  WARN mz_storage_client::controller::rehydration: storage cluster produced error, reconnecting: status: Unknown, message: "error reading a body from connection: broken pipe", details: [], metadata: MetadataMap { headers: {} }
```
Comment (vmarcos) [inline test/sqllogictest/transform/fold_vs_dataflow/5_repeat_row_dataflow.slt:36]: This looks like an instance of https://github.com/MaterializeInc/materialize/issues/18738. The panic should disappear in `release` mode and instead manifest as a wrap-around.

**Review by vmarcos — COMMENTED** [review]
Thanks for working on these tests! Not sure how much time you are willing to invest into refining these, but here are a few more items that would be good to validate:

- Top-K computation, both with LATERAL JOIN (i.e., with groups) or not.
- More coverage of set operations, e.g., `EXCEPT`, `EXCEPT ALL`.
- More coverage of joins.
- If possible, filters.

The constant-folding transformation has arms to handle all of the above: https://github.com/MaterializeInc/materialize/blob/8147c0b1a6a4b6d0c318f947a23ca4c57ad47142/src/transform/src/fold_constants.rs#L85-L402.

A complementary approach to having a fixed test suite that runs as part as our fast SQL logic tests is to also include a set of constant relations (created as views the way you do in the tests in this PR) in our fuzzing nightlies with SQLsmith. Would this be feasible?

**Comment by def-** [comment]
> A complementary approach to having a fixed test suite that runs as part as our fast SQL logic tests is to also include a set of constant relations (created as views the way you do in the tests in this PR) in our fuzzing nightlies with SQLsmith. Would this be feasible?

SQLsmith might be difficult. I think a new oracle for SQLancer would be possible, but some effort and probably on my plate.

**Review by philip-stoev — APPROVED** [review]
I would defer to @vmarcos so say if more testing is needed as part of this task -- if this is the case, I think some sort of scripting should be used in order to exercise the various combinations more exhaustively.

**Comment by vmarcos** [comment]
> I would defer to @vmarcos so say if more testing is needed as part of this task -- if this is the case, I think some sort of scripting should be used in order to exercise the various combinations more exhaustively.

I think that to uncover potential new issues, a scripting approach would be ideal, yes. That way, we'd get more coverage of the constant-folding code, increasing confidence. This PR, however, already improves on our current state of affairs. So would you be OK with saying this PR advances https://github.com/MaterializeInc/materialize/issues/17767, but does not close it? That way, we can take on follow-up work separately and merge this now.

**Comment by nrainer-materialize** [comment]
> I would defer to @vmarcos so say if more testing is needed as part of this task -- if this is the case, I think some sort of scripting should be used in order to exercise the various combinations more exhaustively.

I took the time to briefly sketched a [small Python script to generate such code](https://github.com/nrainer-materialize/materialize/commit/290ba10cadc5114f475147660e252edf5146b3cd) for mathematical operations. An excerpt of the output is:
```

# -------------------------------
# BEGIN TYPE UINT2 USING VIEW (constant folding)
# -------------------------------

statement ok
CREATE VIEW v_data_UINT2 (zero_value, max_value, smallest_value) AS SELECT 0::UINT2, 65535::UINT2, 1::UINT2, 1::UINT2;

query III
SELECT SUM(max_value), MIN(max_value), MAX(max_value) FROM v_data_UINT2;
----
<to be generated with --write-results>

query III
SELECT SUM(max_value + max_value), MIN(max_value + max_value), MAX(max_value + max_value) FROM v_data_UINT2;
----
<to be generated with --write-results>
```

@philip-stoev: Is this something I should proceed with or would you like me to continue on other issues with higher priorities?

**Review by def- — APPROVED** [review]

**Review by vmarcos — APPROVED** [review]
Sounds good to me! It would be great also if you would either create another issue for the follow-up work or not close the original one, whichever you prefer.

**Comment by nrainer-materialize** [comment]
> Sounds good to me! It would be great also if you would either create another issue for the follow-up work or not close the original one, whichever you prefer.

To be addressed with https://github.com/MaterializeInc/materialize/issues/18858.

**Pull request merged**